openmdao/components/subproblem_comp.py

```diff
@@ -0,0 +1,344 @@
+                                       if name in self.coloring._row_vars]
+            self.coloring._col_vars = [meta['prom_name'] for _, meta in boundary_inputs]
+
+        # self.sparsity = self.coloring.get_subjac_sparsity()
+
+        # TODO clean this up with previously defined lists if I can
+        if self.model_input_names is None:
+            self.model_input_names = [meta['prom_name'] for _, meta in
+                                      boundary_inputs]
+            # self.model_input_names = boundary_inputs
+
+        # don't want to include `IndepVarComp`s as outputs
+        if self.model_output_names is None:
+            self.model_output_names = [meta['prom_name'] for _, meta in
+                                       p.model.list_outputs(out_stream=None, prom_name=True,
+                                                            is_indep_var=False)]
+
+        self.options.update(_get_model_vars('inputs', self.model_input_names, boundary_inputs))
+        self.options.update(_get_model_vars('outputs', self.model_output_names, all_outputs))
+
+        inputs = self.options['inputs']
+        outputs = self.options['outputs']
+
+        # instantiate input/output name list for use in compute and
+        # compute partials
+        self._input_names = []
+        self._output_names = []
+
+        # remove the `prom_name` from the metadata and then store it for each
+        # input and output
+        for var, meta in inputs.items():
+            prom_name = meta.pop('prom_name')
+            self.add_input(var, **meta)
+            meta['prom_name'] = prom_name
+            self._input_names.append(var)
+
+        for var, meta in outputs.items():
+            prom_name = meta.pop('prom_name')
+            self.add_output(var, **meta)
+            meta['prom_name'] = prom_name
+            self._output_names.append(var)
+
+        if self.coloring is None:
+            self.declare_partials(of='*', wrt='*')
+        else:
+            for of, wrt, nzrows, nzcols, _, _, _, _ in self.coloring._subjac_sparsity_iter():
+                if of not in self._output_names or wrt not in self._input_names:
+                    continue
+                self.declare_partials(of=of, wrt=wrt, rows=nzrows, cols=nzcols)
+
+    def _set_complex_step_mode(self, active):
+        super()._set_complex_step_mode(active)
+        self._subprob.set_complex_step_mode(active)
+
+    def compute(self, inputs, outputs):
+        """
+        Perform the subproblem system computation at run time.
+
+        Parameters
+        ----------
+        inputs : Vector
+            Unscaled, dimensional input variables read via inputs[key].
+        outputs : Vector
+            Unscaled, dimensional output variables read via outputs[key].
+        """
+        p = self._subprob
+
+        # switch subproblem to use complex IO if in complex step mode
+        self._set_complex_step_mode(self.under_complex_step)
```

Comment:
I'm not clear on why this call to `_set_complex_step_mode` is necessary.  Wouldn't we already be in the right state since we have overridden the `_set_complex_step_mode` method?